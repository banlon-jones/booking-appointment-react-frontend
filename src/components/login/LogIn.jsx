import axios from 'axios';
import React, { useState, useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { Button, Form, Container } from 'react-bootstrap';
import { useNavigate } from 'react-router-dom';
import classes from './LogIn.module.css';
import { logInUser, setRole } from '../../store/user/user';
import CircleSpinner from '../spinners/CircleSpinner';

const LogIn = () => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [errorMessage, setErrorMessage] = useState('');
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    setErrorMessage('');
  }, [email, password]);

  const logInButtonContent = () => {
    if (!loading) {
      return 'Log In';
    }
    return <CircleSpinner />;
  };

  const loginUser = async (e) => {
    e.preventDefault();
    try {
      setLoading(true);
      const { data } = await axios({
        url: 'https://resorts-booking-api.herokuapp.com/auth/login',
        method: 'POST',
        data: {
          email,
          password,
        },
      });
      if (data.error) {
        setErrorMessage(data.error);
        return;
      }
      sessionStorage.setItem('JwtAccessToken', data);
      dispatch(logInUser());
      dispatch(setRole());
      navigate('/');
    } catch (err) {
      setLoading(false);
      const apiErrorMessages = {
        emailErr: 'User with provided email not found!',
        passwordErr: 'Incorrect password provided!',
      };
      const { error } = err.response.data;
      if (error.match("Couldn't find User with...")) {
        setErrorMessage(apiErrorMessages.emailErr);
      } else {
        setErrorMessage(apiErrorMessages.passwordErr);
      }
    }
  };

  return (
    <div className={classes.login_container}>
      <Container className={classes.form_container}>
        <h3 className={classes.form_title}>Log In</h3>
        <span className={classes.error_message}>{errorMessage}</span>
        <Form onSubmit={(e) => loginUser(e)}>
          <Form.Group className="mb-3" controlId="formBasicEmail">
            <Form.Label>Email address</Form.Label>
            <span>*</span>
            <Form.Control
              type="email"
              placeholder="Enter email"
              onChange={(e) => setEmail(e.target.value)}
              required
            />
          </Form.Group>
          <Form.Group className="mb-3" controlId="formBasicPassword">
            <Form.Label>Password</Form.Label>
            <span>*</span>
            <Form.Control
              type="password"
              placeholder="Password"
              onChange={(e) => setPassword(e.target.value)}
              required
            />
          </Form.Group>
          <Button
            variant="primary"
            type="submit"
            className={classes.submit_btn}
          >
            {logInButtonContent()}
          </Button>
        </Form>
      </Container>
    </div>
  );
};

export default LogIn;
