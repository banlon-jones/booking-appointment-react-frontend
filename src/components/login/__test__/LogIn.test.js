import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { BrowserRouter as Router } from 'react-router-dom';
import LogIn from '../LogIn';

describe('snapshots', () => {
  const initialState = { user: { loggedIn: false, role: '' } };
  const mockStore = configureStore();
  let store;

  it('matches LogIn snapshot', () => {
    store = mockStore(initialState);
    const tree = renderer
      .create(
        <Provider store={store}>
          <Router>
            <LogIn />
          </Router>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
