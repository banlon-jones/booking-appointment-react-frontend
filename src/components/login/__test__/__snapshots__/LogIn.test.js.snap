// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots matches LogIn snapshot 1`] = `
<div
  className="login_container"
  data-testid="login-cont"
>
  <div
    className="form_container container"
  >
    <h3
      className="form_title"
    >
      Log In
    </h3>
    <span
      className="error_message"
    />
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicEmail"
        >
          Email address
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicEmail"
          onChange={[Function]}
          placeholder="Enter email"
          required={true}
          type="email"
        />
      </div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicPassword"
        >
          Password
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicPassword"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
        />
      </div>
      <button
        className="submit_btn btn btn-primary"
        disabled={false}
        type="submit"
      >
        Log In
      </button>
    </form>
  </div>
</div>
`;

exports[`snapshots matches snapshot 1`] = `
<div
  className="login_container"
  data-testid="login-cont"
>
  <div
    className="form_container container"
  >
    <h3
      className="form_title"
    >
      Log In
    </h3>
    <span
      className="error_message"
    />
    <form
      className=""
      onSubmit={[Function]}
    >
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicEmail"
        >
          Email address
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicEmail"
          onChange={[Function]}
          placeholder="Enter email"
          required={true}
          type="email"
        />
      </div>
      <div
        className="mb-3"
      >
        <label
          className="form-label"
          htmlFor="formBasicPassword"
        >
          Password
        </label>
        <span>
          *
        </span>
        <input
          className="form-control"
          id="formBasicPassword"
          onChange={[Function]}
          placeholder="Password"
          required={true}
          type="password"
        />
      </div>
      <button
        className="submit_btn btn btn-primary"
        disabled={false}
        type="submit"
      >
        Log In
      </button>
    </form>
  </div>
</div>
`;
